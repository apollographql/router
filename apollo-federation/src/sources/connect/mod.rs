#![allow(unused_imports)]

mod json_selection;
mod url_path_template;

pub use json_selection::ApplyTo;
pub use json_selection::ApplyToError;
pub use json_selection::JSONSelection;
pub use json_selection::Key;
pub use json_selection::PathSelection;
pub use json_selection::SubSelection;
pub use url_path_template::URLPathTemplate;
