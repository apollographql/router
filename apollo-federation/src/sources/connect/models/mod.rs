mod validation;

use apollo_compiler::NodeStr;
use indexmap::IndexMap;

use crate::{
    error::FederationError,
    schema::ValidFederationSchema,
    sources::connect::{
        spec::{extract_connect_directive_arguments, extract_source_directive_arguments},
        ConnectSpecDefinition,
    },
};

use super::{
    spec::{ConnectHTTPArguments, HTTPHeaderOption, SourceHTTPArguments},
    ConnectId, Selection, URLPathTemplate,
};

// --- Connector ---------------------------------------------------------------

#[cfg_attr(test, derive(Debug))]
pub(crate) struct Connector {
    pub(crate) id: ConnectId,
    transport: Transport,
    pub(crate) selection: Selection,
}

#[cfg_attr(test, derive(Debug))]
enum Transport {
    HttpJson(HttpJsonTransport),
}

impl Connector {
    pub(crate) fn from_valid_schema(
        schema: &ValidFederationSchema,
        subgraph_name: NodeStr,
    ) -> Result<IndexMap<ConnectId, Self>, FederationError> {
        let Some(metadata) = schema.metadata() else {
            return Ok(IndexMap::new());
        };

        let Some(link) = metadata.for_identity(&ConnectSpecDefinition::identity()) else {
            return Ok(IndexMap::new());
        };

        let source_name = ConnectSpecDefinition::source_directive_name(&link);
        let source_arguments = extract_source_directive_arguments(schema, &source_name)?;

        let connect_name = ConnectSpecDefinition::connect_directive_name(&link);
        let connect_arguments = extract_connect_directive_arguments(schema, &connect_name)?;

        connect_arguments
            .into_iter()
            .map(move |args| {
                let id = ConnectId {
                    subgraph_name: subgraph_name.clone(),
                    directive: args.position,
                };

                let source = if let Some(source_name) = args.source {
                    source_arguments
                        .iter()
                        .find(|source| source.name == source_name)
                } else {
                    None
                };

                let connect_http = args.http.expect("@connect http missing");
                let source_http = source.map(|s| &s.http);

                let transport = Transport::HttpJson(HttpJsonTransport::from_directive(
                    &connect_http,
                    source_http,
                )?);

                Ok((
                    id.clone(),
                    Connector {
                        id,
                        transport,
                        selection: args.selection,
                    },
                ))
            })
            .collect()
    }
}

// --- HTTP JSON ---------------------------------------------------------------

#[cfg_attr(test, derive(Debug))]
struct HttpJsonTransport {
    base_url: NodeStr,
    path_template: URLPathTemplate,
    method: HTTPMethod,
    headers: IndexMap<NodeStr, Option<HTTPHeaderOption>>,
    body: Option<Selection>,
}

impl HttpJsonTransport {
    fn from_directive(
        http: &ConnectHTTPArguments,
        source: Option<&SourceHTTPArguments>,
    ) -> Result<Self, FederationError> {
        let (method, path) = if let Some(path) = &http.get {
            (HTTPMethod::Get, path)
        } else if let Some(path) = &http.post {
            (HTTPMethod::Post, path)
        } else if let Some(path) = &http.patch {
            (HTTPMethod::Patch, path)
        } else if let Some(path) = &http.put {
            (HTTPMethod::Put, path)
        } else if let Some(path) = &http.delete {
            (HTTPMethod::Delete, path)
        } else {
            return Err(FederationError::internal("missing http method"));
        };

        let base_url = if let Some(base_url) = source.as_ref().map(|source| source.base_url.clone())
        {
            base_url
        } else {
            todo!("get base url from path")
        };

        let mut headers = source
            .as_ref()
            .map(|source| source.headers.0.clone())
            .unwrap_or_default();
        headers.extend(http.headers.0.clone());

        Ok(Self {
            base_url,
            path_template: URLPathTemplate::parse(path).expect("path template"),
            method,
            headers,
            body: http.body.clone(),
        })
    }
}

/// The HTTP arguments needed for a connect request
#[cfg_attr(test, derive(Debug))]
pub(crate) enum HTTPMethod {
    Get,
    Post,
    Patch,
    Put,
    Delete,
}

#[cfg(test)]
mod tests {
    use apollo_compiler::Schema;

    use crate::{
        query_graph::extract_subgraphs_from_supergraph::extract_subgraphs_from_supergraph,
        schema::FederationSchema, ValidFederationSubgraphs,
    };

    use super::*;

    static SIMPLE_SUPERGRAPH: &str = include_str!("../tests/schemas/simple.graphql");

    fn get_subgraphs(supergraph_sdl: &str) -> ValidFederationSubgraphs {
        let schema = Schema::parse(supergraph_sdl, "supergraph.graphql").unwrap();
        let supergraph_schema = FederationSchema::new(schema).unwrap();
        extract_subgraphs_from_supergraph(&supergraph_schema, Some(true)).unwrap()
    }

    #[test]
    fn test_from_schema() {
        let subgraphs = get_subgraphs(SIMPLE_SUPERGRAPH);
        let subgraph = subgraphs.get("connectors").unwrap();
        let connectors =
            Connector::from_valid_schema(&subgraph.schema, "connectors".into()).unwrap();
        dbg!(&connectors);
    }
}
