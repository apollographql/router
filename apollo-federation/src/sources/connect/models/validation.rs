//! Validation of the `@source` and `@connect` directives.

// No panics allowed in this module
#![cfg_attr(
    not(test),
    deny(
        clippy::exit,
        clippy::panic,
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::indexing_slicing,
        clippy::unimplemented,
        clippy::todo
    )
)]

/* TODO:

## @connect

- Disallowed on Subscription fields
- http:
  - Has one path (GET, POST, etc)
  - URL path template valid syntax
  - body is valid syntax
- selection
  - Valid syntax
- If source:
  - Matches a @source
  - Transport is the same as the @source
- If no source:
  - http:
    - Path is a fully qualified URL (plus a URL path template)
- If entity: true
  - Must be on Query type
  - Field arguments must match fields in type (and @key if present)
  - Required for types with @key (eventually with @finder)

## HTTPHeaderMapping

- name: is unique
- name: is a valid header name
- as: is a valid header name
- value: is a list of valid header values
- as: and value: cannot both be present

## Output selection

- Empty composite selection
- Leaf with selections

## Input selections (URL path templates and bodies)

- Missing $args
- Missing $this

## @link

- Rename @connect import

*/

use std::collections::HashMap;
use std::error::Error;
use std::fmt::Display;

use apollo_compiler::ast::Value;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use apollo_compiler::Node;
use apollo_compiler::NodeLocation;
use apollo_compiler::Schema;
use url::Url;

use crate::link::Link;
use crate::sources::connect::spec::schema::SOURCE_BASE_URL_ARGUMENT_NAME;
use crate::sources::connect::spec::schema::SOURCE_HTTP_ARGUMENT_NAME;
use crate::sources::connect::spec::schema::SOURCE_NAME_ARGUMENT_NAME;
use crate::sources::connect::ConnectSpecDefinition;

/// Validate the connectors-related directives `@source` and `@connect`.
///
/// This function attempts to collect as many validation errors as possible, so it does not bail
/// out as soon as it encounters one.
pub fn validate(schema: Schema) -> Vec<ValidationError> {
    let connect_identity = ConnectSpecDefinition::identity();
    let link = schema
        .schema_definition
        .directives
        .iter()
        .find_map(|directive| {
            let link = Link::from_directive_application(directive).ok()?;
            if link.url.identity == connect_identity {
                Some(link)
            } else {
                None
            }
        });
    let Some(link) = link else {
        return vec![]; // There are no connectors-related directives to validate
    };
    let source_directive_name = ConnectSpecDefinition::source_directive_name(&link);
    let source_directives: Vec<SourceDirective> = schema
        .schema_definition
        .directives
        .iter()
        .filter(|directive| directive.name == source_directive_name)
        .map(validate_source)
        .collect();

    let mut errors = Vec::new();
    let mut names = HashMap::new();
    for directive in source_directives {
        errors.extend(directive.errors);
        match directive.name.into_value_or_error() {
            Err(error) => errors.push(error),
            Ok(name) => names
                .entry(name)
                .or_insert_with(Vec::new)
                .extend(directive.directive.node.location().into_iter()),
        }
    }
    for (name, locations) in names {
        if locations.len() > 1 {
            errors.push(ValidationError {
                message: format!("every @{source_directive_name} name must be unique; found duplicate source name {name}"),
                code: ErrorCode::DuplicateSourceName,
                locations,
            });
        }
    }
    errors
}

fn validate_source(directive: &Component<Directive>) -> SourceDirective {
    let name = SourceName::from_directive(directive);
    let url_error = directive
        .arguments
        .iter()
        .find(|arg| arg.name == SOURCE_HTTP_ARGUMENT_NAME)
        .and_then(|arg| {
            let value = arg
                .value
                .as_object()?
                .iter()
                .find_map(|(name, value)| {
                    (*name == SOURCE_BASE_URL_ARGUMENT_NAME).then_some(value)
                })?
                .as_str()?;
            let url = match Url::parse(value) {
                Err(inner) => {
                    return Some(ValidationError {
                        code: ErrorCode::SourceUrl,
                        message: format!(
                            "baseURL argument for {name} was not a valid URL: {inner}"
                        ),
                        locations: arg.location().into_iter().collect(),
                    })
                }
                Ok(url) => url,
            };
            let scheme = url.scheme();
            if scheme != "http" && scheme != "https" {
                return Some(ValidationError {
                    code: ErrorCode::SourceScheme,
                    message: format!(
                        "baseURL argument for {name} must be http or https, got {scheme}"
                    ),
                    locations: arg.value.location().into_iter().collect(),
                });
            }
            None
        });
    let mut errors = Vec::new();
    if let Some(url_error) = url_error {
        errors.push(url_error);
    }
    SourceDirective {
        name,
        errors,
        directive: directive.clone(),
    }
}

/// A `@source` directive along with any errors related to it.
struct SourceDirective {
    name: SourceName,
    errors: Vec<ValidationError>,
    directive: Component<Directive>,
}

/// The `name` argument of a `@source` directive.
#[derive(Clone, Debug)]
enum SourceName {
    /// A perfectly reasonable source name.
    Valid {
        value: Node<Value>,
        directive_name: DirectiveName,
    },
    /// Contains invalid characters, so it will have to be renamed. This means certain checks
    /// (like uniqueness) should be skipped. However, we have _a_ name, so _other_ checks on the
    /// `@source` directive can continue.
    Invalid {
        value: Node<Value>,
        directive_name: DirectiveName,
    },
    /// The name was an empty string
    Empty {
        directive_name: DirectiveName,
        value: Node<Value>,
    },
    /// No `name` argument was defined
    Missing {
        directive_name: DirectiveName,
        ast_node: Node<Directive>,
    },
}

type DirectiveName = String;

impl SourceName {
    fn from_directive(directive: &Component<Directive>) -> Self {
        let directive_name = directive.name.to_string();
        let Some(arg) = directive
            .arguments
            .iter()
            .find(|arg| arg.name == SOURCE_NAME_ARGUMENT_NAME)
        else {
            return Self::Missing {
                directive_name,
                ast_node: directive.node.clone(),
            };
        };
        let Some(str_value) = arg.value.as_str() else {
            return Self::Invalid {
                value: arg.value.clone(),
                directive_name,
            };
        };
        if str_value.is_empty() {
            Self::Empty {
                directive_name,
                value: arg.value.clone(),
            }
        } else if str_value.chars().all(|c| c.is_alphanumeric() || c == '_') {
            Self::Valid {
                value: arg.value.clone(),
                directive_name,
            }
        } else {
            Self::Invalid {
                value: arg.value.clone(),
                directive_name,
            }
        }
    }

    pub fn into_value_or_error(self) -> Result<Node<Value>, ValidationError> {
        match self {
            Self::Valid { value, ..} => Ok(value),
            Self::Invalid {
                value,
                directive_name,
            } => Err(ValidationError {
                message: format!("invalid characters in @{directive_name} name {value}, only alphanumeric and underscores are allowed"),
                code: ErrorCode::InvalidSourceName,
                locations: value.location().into_iter().collect(),
            }),
            Self::Empty { directive_name, value } => {
                Err(ValidationError {
                    code: ErrorCode::EmptySourceName,
                    message: format!("name argument to @{directive_name} can't be empty"),
                    locations: value.location().into_iter().collect(),
                })
            }
            Self::Missing { directive_name, ast_node } => Err(ValidationError {
                code: ErrorCode::GraphQLError,
                message: format!("missing name argument to @{directive_name}"),
                locations: ast_node.location().into_iter().collect()
            }),
        }
    }
}

impl Display for SourceName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Valid {
                value,
                directive_name,
            }
            | Self::Invalid {
                value,
                directive_name,
            } => write!(f, "@{directive_name} {value}"),
            Self::Empty { directive_name, .. } | Self::Missing { directive_name, .. } => {
                write!(f, "unnamed @{directive_name}")
            }
        }
    }
}

#[derive(Debug)]
pub struct ValidationError {
    pub code: ErrorCode,
    pub message: String,
    pub locations: Vec<NodeLocation>,
}

impl Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl Error for ValidationError {}

#[derive(Clone, Copy, Debug)]
pub enum ErrorCode {
    GraphQLError,
    DuplicateSourceName,
    InvalidSourceName,
    EmptySourceName,
    SourceUrl,
    SourceScheme,
}

#[cfg(test)]
mod test_validate_source {
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn unparseable_url() {
        let schema = include_str!("test_data/invalid_source_url.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }

    #[test]
    fn incorrect_scheme() {
        let schema = include_str!("test_data/invalid_source_url_scheme.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }

    #[test]
    fn invalid_chars_in_source_name() {
        let schema = include_str!("test_data/invalid_chars_in_source_name.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }

    #[test]
    fn empty_source_name() {
        let schema = include_str!("test_data/empty_source_name.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }

    #[test]
    fn duplicate_source_name() {
        let schema = include_str!("test_data/duplicate_source_name.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }

    #[test]
    fn multiple_errors() {
        let schema = include_str!("test_data/multiple_errors.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 2);
        assert_snapshot!(errors[0].to_string());
        assert_snapshot!(errors[1].to_string());
    }

    #[test]
    fn source_directive_rename() {
        let schema = include_str!("test_data/source_directive_rename.graphql");
        let schema = Schema::parse(schema, "test.graphql").unwrap();
        let errors = validate(schema);
        assert_eq!(errors.len(), 1);
        assert_snapshot!(errors[0].to_string());
    }
}
