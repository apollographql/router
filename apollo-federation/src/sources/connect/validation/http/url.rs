use std::fmt::Display;
use std::str::FromStr;

use apollo_compiler::ast::FieldDefinition;
use apollo_compiler::ast::Type;
use apollo_compiler::ast::Value;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::Node;
use apollo_compiler::Schema;
use url::Url;

use crate::sources::connect::url_template;
use crate::sources::connect::url_template::VariableType;
use crate::sources::connect::validation::coordinates::HttpMethodCoordinate;
use crate::sources::connect::validation::graphql::GraphQLString;
use crate::sources::connect::validation::graphql::SchemaInfo;
use crate::sources::connect::validation::Code;
use crate::sources::connect::validation::Message;
use crate::sources::connect::URLTemplate;
use crate::sources::connect::Variable;

pub(crate) fn validate_template(
    coordinate: HttpMethodCoordinate,
    schema: &SchemaInfo,
) -> Result<URLTemplate, Vec<Message>> {
    let (template, str_value) = match parse_template(coordinate, schema) {
        Ok(tuple) => tuple,
        Err(message) => return Err(vec![message]),
    };
    let mut messages = Vec::new();
    if let Some(base) = template.base.as_ref() {
        messages
            .extend(validate_base_url(base, coordinate, coordinate.node, str_value, schema).err());
    }

    for variable in template.path_variables() {
        match validate_variable(variable, str_value, coordinate, schema) {
            Err(err) => messages.push(err),
            Ok(Some(ty)) if !ty.is_non_null() => {
                messages.push(Message {
                    code: Code::NullablePathVariable,
                    message: format!(
                        "Variables in path parameters should be non-null, but {coordinate} contains `{{{variable}}}` which is nullable. \
                         If a null value is provided at runtime, the request will fail.",
                    ),
                    locations: str_value.line_col_for_subslice(
                        variable.location.clone(),
                        schema,
                    ).into_iter().collect(),
                });
            }
            Ok(_) => {} // Type is non-null, or unknowable
        }
    }

    for variable in template.query_variables() {
        if let Err(err) = validate_variable(variable, str_value, coordinate, schema) {
            messages.push(err);
        }
    }

    if messages.is_empty() {
        Ok(template)
    } else {
        Err(messages)
    }
}

fn parse_template<'schema>(
    coordinate: HttpMethodCoordinate<'schema>,
    schema: &'schema SchemaInfo,
) -> Result<(URLTemplate, GraphQLString<'schema>), Message> {
    let str_value = GraphQLString::new(coordinate.node, &schema.sources).map_err(|_| Message {
        code: Code::GraphQLError,
        message: format!("The value for {coordinate} must be a string."),
        locations: coordinate
            .node
            .line_column_range(&schema.sources)
            .into_iter()
            .collect(),
    })?;
    let template = URLTemplate::from_str(str_value.as_str()).map_err(
        |url_template::Error { message, location }| Message {
            code: Code::InvalidUrl,
            message: format!("{coordinate} must be a valid URL template. {message}"),
            locations: location
                .and_then(|location| str_value.line_col_for_subslice(location, schema))
                .into_iter()
                .collect(),
        },
    )?;
    Ok((template, str_value))
}

pub(crate) fn validate_base_url(
    url: &Url,
    coordinate: impl Display,
    value: &Node<Value>,
    str_value: GraphQLString,
    schema: &SchemaInfo,
) -> Result<(), Message> {
    let scheme = url.scheme();
    if scheme != "http" && scheme != "https" {
        let scheme_location = 0..scheme.len();
        Err(Message {
            code: Code::InvalidUrlScheme,
            message: format!(
                "The value {value} for {coordinate} must be http or https, got {scheme}.",
            ),
            locations: str_value
                .line_col_for_subslice(scheme_location, schema)
                .into_iter()
                .collect(),
        })
    } else {
        Ok(())
    }
}

fn validate_variable<'schema>(
    variable: &'schema Variable,
    url_value: GraphQLString,
    coordinate: HttpMethodCoordinate<'schema>,
    schema: &'schema SchemaInfo,
) -> Result<Option<Type>, Message> {
    let field_coordinate = coordinate.connect.field_coordinate;
    let field = field_coordinate.field;
    let mut path = variable.path.split('.');
    let path_root = path.next().unwrap_or(&variable.path);
    let mut path_component_start = variable.location.start + variable.var_type.as_str().len() + 1;
    let mut path_component_end = path_component_start + path_root.len();
    let mut variable_type = match variable.var_type {
        VariableType::Config => {
            return Ok(None); // We don't validate Router config yet
        }
        VariableType::Args => {
            field.arguments.iter().find(|arg| arg.name == path_root).ok_or_else( || {
                Message {
                    code: Code::UndefinedArgument,
                    message: format!(
                        "{coordinate} contains `{{{variable}}}`, but {field_coordinate} does not have an argument named `{path_root}`.",
                    ),
                    locations: url_value.line_col_for_subslice(
                        path_component_start..path_component_end,
                        schema,
                    ).into_iter().collect()
                }
            }).map(|arg| arg.ty.as_ref().clone())?
        }
        VariableType::This => {
            field_coordinate.object.fields.get(path_root).ok_or_else(||Message {
                    code: Code::UndefinedField,
                    message: format!(
                        "{coordinate} contains `{{{variable}}}`, but {object} does not have a field named `{path_root}`.",
                        object = field_coordinate.object.name,
                    ),
                    locations: url_value.line_col_for_subslice(
                        path_component_start..path_component_end,
                        schema,
                    ).into_iter().collect()
                }).map(|field| field.ty.clone())?
        }
    };

    for nested_field_name in path {
        path_component_start = path_component_end + 1; // Past the last component and its dot
        path_component_end = path_component_start + nested_field_name.len();
        let parent_is_nullable = !variable_type.is_non_null();
        variable_type = resolve_type(schema, &variable_type, field_coordinate.field)
            .and_then(|extended_type| {
                match extended_type {
                    ExtendedType::Enum(_) | ExtendedType::Scalar(_) => None,
                    ExtendedType::Object(object) => object.fields.get(nested_field_name).map(|field| &field.ty),
                    ExtendedType::InputObject(input_object) => input_object.fields.get(nested_field_name).map(|field| field.ty.as_ref()),
                    // TODO: at the time of writing, you can't declare interfaces or unions in connectors schemas at all, so these aren't tested
                    ExtendedType::Interface(interface) => interface.fields.get(nested_field_name).map(|field| &field.ty),
                    ExtendedType::Union(_) => {
                        return Err(Message {
                            code: Code::UnsupportedVariableType,
                            message: format!(
                                "The type {variable_type} is a union, which is not supported in variables yet.",
                            ),
                            locations: field_coordinate
                                .field
                                .line_column_range(&schema.sources)
                                .into_iter()
                                .collect(),
                        })
                    },
                }
                    .ok_or_else(|| Message {
                        code: Code::UndefinedField,
                        message: format!(
                            "{coordinate} contains `{{{variable}}}`, but `{variable_type}` does not have a field named `{nested_field_name}`.",
                        ),
                        locations: url_value.line_col_for_subslice(
                            path_component_start..path_component_end,
                            schema,
                        ).into_iter().collect(),
                    })
            })?.clone();
        if parent_is_nullable && variable_type.is_non_null() {
            variable_type = variable_type.nullable();
        }
    }

    Ok(Some(variable_type))
}

fn resolve_type<'schema>(
    schema: &'schema Schema,
    ty: &Type,
    definition: &Component<FieldDefinition>,
) -> Result<&'schema ExtendedType, Message> {
    schema
        .types
        .get(ty.inner_named_type())
        .ok_or_else(|| Message {
            code: Code::GraphQLError,
            message: format!("The type {ty} is referenced but not defined in the schema.",),
            locations: definition
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
}
