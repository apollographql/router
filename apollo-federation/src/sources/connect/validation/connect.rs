//! Parsing and validation of `@connect` directives

use apollo_compiler::Name;
use apollo_compiler::Node;
use apollo_compiler::ast::FieldDefinition;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::ObjectType;
use itertools::Itertools;

use self::entity::validate_entity_arg;
use self::selection::Selection;
use super::Code;
use super::Message;
use super::coordinates::ConnectDirectiveCoordinate;
use super::coordinates::connect_directive_name_coordinate;
use super::coordinates::source_name_value_coordinate;
use super::source::SourceName;
use crate::sources::connect::ConnectSpec;
use crate::sources::connect::id::ConnectedElement;
use crate::sources::connect::id::ObjectCategory;
use crate::sources::connect::spec::schema::CONNECT_SOURCE_ARGUMENT_NAME;
use crate::sources::connect::validation::connect::http::Http;
use crate::sources::connect::validation::graphql::SchemaInfo;

mod entity;
mod http;
mod selection;

pub(super) fn fields_seen_by_all_connects(
    schema: &SchemaInfo,
    all_source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    schema
        .types
        .values()
        .filter(|ty| !ty.is_built_in())
        .for_each(|extended_type| {
            if let ExtendedType::Object(node) = extended_type {
                match fields_seen_by_connectors_on_types(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }

                match fields_seen_by_connectors_on_fields(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }
            }
        });

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on types are valid
fn fields_seen_by_connectors_on_types(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let connect_directives = object
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // TODO: find a better place for feature gates like this
    if schema.connect_link.spec == ConnectSpec::V0_1 {
        return Err(vec![Message {
            code: Code::FeatureUnavailable,
            message: format!(
                "Using `@{connect_directive_name}` on `type {object_name}` requires connectors v0.2. Learn more at https://go.apollo.dev/connectors/changelog.",
                object_name = object.name,
                connect_directive_name = schema.connect_directive_name(),
            ),
            locations: object
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        }]);
    }

    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Type { type_def: object },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                messages.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => messages.push(error),
        }

        // TODO: validate batch/this arguments as key fields

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    messages.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                messages.extend(errs);
                continue;
            }
        };
        messages.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on object fields are valid
fn fields_seen_by_connectors_on_fields(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let object_category = if schema
        .schema_definition
        .query
        .as_ref()
        .is_some_and(|query| query.name == object.name)
    {
        ObjectCategory::Query
    } else if schema
        .schema_definition
        .mutation
        .as_ref()
        .is_some_and(|mutation| mutation.name == object.name)
    {
        ObjectCategory::Mutation
    } else {
        ObjectCategory::Other
    };
    let mut seen_fields = Vec::new();
    let mut messages = Vec::new();
    for field in object.fields.values() {
        match fields_seen_by_connector(field, object_category, source_names, object, schema) {
            Ok(fields) => seen_fields.extend(fields),
            Err(errs) => messages.extend(errs),
        }
    }
    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

fn fields_seen_by_connector(
    field: &Component<FieldDefinition>,
    category: ObjectCategory,
    source_names: &[SourceName],
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let source_map = &schema.sources;
    let mut errors = Vec::new();
    let connect_directives = field
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // mark the field with a @connect directive as seen
    let mut seen_fields = vec![(object.name.clone(), field.name.clone())];

    // direct recursion isn't allowed, like a connector on User.friends: [User]
    if &object.name == field.ty.inner_named_type() {
        errors.push(Message {
            code: Code::CircularReference,
            message: format!(
                "Direct circular reference detected in `{}.{}: {}`. For more information, see https://go.apollo.dev/connectors/limitations#circular-references",
                object.name,
                field.name,
                field.ty
            ),
            locations: field.line_column_range(source_map).into_iter().collect(),
        });
    }

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Field {
                parent_type: object,
                parent_category: category,
                field_def: field,
            },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                errors.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => errors.push(error),
        }

        errors
            .extend(validate_entity_arg(field, connect_directive, object, schema, category).err());

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    errors.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                errors.extend(errs);
                continue;
            }
        };
        errors.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if errors.is_empty() {
        Ok(seen_fields)
    } else {
        Err(errors)
    }
}

pub(super) fn validate_source_name<'schema>(
    directive: &Node<Directive>,
    coordinate: &ConnectDirectiveCoordinate,
    source_names: &'schema [SourceName],
    schema: &SchemaInfo,
) -> Result<Option<&'schema SourceName<'schema>>, Message> {
    let Some(source_name_arg) = directive
        .arguments
        .iter()
        .find(|arg| arg.name == CONNECT_SOURCE_ARGUMENT_NAME)
    else {
        return Ok(None);
    };

    let resolved_source_name = source_names
        .iter()
        .find(|name| **name == source_name_arg.value);

    if let Some(source_name) = resolved_source_name {
        return Ok(Some(source_name));
    }
    // A source name was set but doesn't match a defined source
    // TODO: Pick a suggestion that's not just the first defined source
    let qualified_directive = connect_directive_name_coordinate(
        schema.connect_directive_name(),
        &source_name_arg.value,
        coordinate,
    );
    if let Some(first_source_name) = source_names.first() {
        Err(Message {
            code: Code::SourceNameMismatch,
            message: format!(
                "{qualified_directive} does not match any defined sources. Did you mean \"{first_source_name}\"?",
                first_source_name = first_source_name.as_str(),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    } else {
        Err(Message {
            code: Code::NoSourcesDefined,
            message: format!(
                "{qualified_directive} specifies a source, but none are defined. Try adding {coordinate} to the schema.",
                coordinate = source_name_value_coordinate(
                    schema.source_directive_name(),
                    &source_name_arg.value
                ),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    }
}
