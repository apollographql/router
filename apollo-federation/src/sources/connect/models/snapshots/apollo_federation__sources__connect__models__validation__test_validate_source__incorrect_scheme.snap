---
source: apollo-federation/src/sources/connect/models/validation.rs
expression: "errors[0].to_string()"
---
baseURL argument for @source "v1" must be http or https, got file
